### PR TITLE
Speed Limit tool - Lane highlighting

### LINKED ISSUE #682: Add lane highlight to Speed Limits tool
When user is hovering a speed limit icon, particularly in lane-wise mode which allows per-lane speed setting, it is sometimes unclear which lane it corresponds to.

image  image

Could we pimp the relevant lane with a highlight to make it completely clear which lane is being altered?

The lane highlight would only be visible while the speed limit icon is hovered.

Suggestion: White highlight (like outgoing lanes on Lane Connector tool) but without the border -a subtle highlight might be sufficient.

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #682 

highlight the affected lanes when user hovers speed limit sprite. 
- highlight  color is blue. when mouse button is pressed the highlight is emboldened and becomes orange.
- If in per lane mode one lane is highlighted. holding shift highlights that lane on the entirety of the road.
- if in directional mode EDIT:all lanes in that direction are highlighted. they highlights are much thicker so that they would merge together as one. holding shift will highlight all affected lanes on that road.  appropriate side of the segment is highlighted.

Screenshot (527)
Cities_9dZikiSpXP

EDIT: I made major changes:
- ~~`TrafficManagerTool` has a new method to render only one side of a segment see screenshots~~
- restored priority sings tool dust bin sprite from old TMPE. I renamed it to clear.png and  put it in a central place to be accessed from multiple places. (#77)
- added new button to restore speed limits to default value. it uses clear.png sprite.
- created back-end to reset speed limit to default state.
- the UI panel has a new toggle to apply changes to entire road. This was handled with shift modifier (which still works)
- pressing shift will toggle "entire road" mode. Pressing the toggle has the same effect. if the toggle is active, shift will deactivate multi-lane mode. When user pressed/releases shift, the toggle gets checked/unchecked to give visual feedback.
- `ShowPerLane` toggle works in the same way as "entire road" toggle but with CTRL key.
- The text of the toggles show the modifier key.

EDIT: I forgot to mention I introduced a couple methods/properties to Lane/Sagement traverse data structures and put some commonly used code in there.

Test: 
- I tested both with LHT and RHT
- I tried swaping start and end node of segments
- I tired inverting the segments.

Screenshot (571)

### CODE REVIEW

**Comment by aubergine10** [comment]
It went a bit bonkers on LHT map (mouse was over the icon at top of pic, but lanes at bottom of pic were highlighted):

image

**Comment by aubergine10** [comment]
Interestingly, whern I hold down Ctrl key for lane-mode, it was highlighting correct lanes.

Seems the inversion issue is only affecting direction-mode.

**Comment by aubergine10** [comment]
In cases where two icons are hovered at same time in lane-mode, only one lane gets highligted:

image

**Comment by aubergine10** [comment]
I suspect this is due to direction inversion issue on LHT maps - it's one way roundabout, in direction mode:

image

**Comment by kianzarrin** [comment]
I replaced m_direction with m_finalDirection. This is fixed now.

I need to develop a test check list because I keep forgetting to test in all possible circumstances.

File: TLM/TLM/UI/Helpers/SegmentLaneMarker.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@ internal class SegmentLaneMarker {
         }
 
         internal Bezier3 Bezier;
+        internal float size = 1.1f;
```
Comment (kvakvs) [inline TLM/TLM/UI/Helpers/SegmentLaneMarker.cs]: is this a const? Can it change?

File: TLM/TLM/UI/SubTools/SpeedLimits/SpeedLimitsTool.cs [hunk 2]
```diff
@@ -200,12 +312,15 @@ public SpeedLimitsTool(TrafficManagerTool mainTool)
                 }
 
                 // no speed limit overlay on selected segment when in vehicle restrictions mode
-                DrawSpeedLimitHandles(
+                hover|= DrawSpeedLimitHandles(
```
Comment (kvakvs) [inline TLM/TLM/UI/SubTools/SpeedLimits/SpeedLimitsTool.cs]: Missing a space.

File: TLM/TLM/UI/Helpers/SegmentLaneMarker.cs (hunk 1 above)
Comment (kianzarrin) [reply TLM/TLM/UI/Helpers/SegmentLaneMarker.cs]: it does change.

**Comment by kianzarrin** [comment]
> @kvakvs  The amount of if (something) conditions is getting out of hand.

"cyclomatic complexity" ... interesting concept.

how should I do this though? move branches into sub methods? 

Note that when the user holds shift, FPS drops because the code goes though a big ass slow for loop. as an optimization effort I created a separate code path for when shift is not pressed so that in most cases FPS is high. I haven't performed any benchmark though. I could do some caching too but the FPS seems fine on my slow computer so I didn't bother.

**Comment by kvakvs** [comment]
If you later revisit this code, yes, please split logic into functions and reduce the nesting.
You will notice that a lot of values are carried between calls and those represent state of this logic flow and maybe deserve their own storage class or struct. The resulting code will be simpler and that is good.

**Comment by aubergine10** [comment]
There's another performance hit with speed limits -- and maybe other tools, haven't checked -- in that to facilitate user being able to 'drag a speed' along a series of segments... this: #388 (not relevant to this PR but something to consider for future performance tuning)

**Comment by aubergine10** [comment]
Need to do some more testing on this; found some issues with parking lane higlight in other PR and need to see if issues also apply to speed limits.

**Comment by aubergine10** [comment]
Retested, only one minor issue found - if two signs hovered at same time, only one lane is highlighting (yet both lanes would be updated if I clicked):

image

**Comment by kianzarrin** [comment]
improved highlighting of segment side:

Screenshot (528)

Cities_UGnW8PubnV

**Comment by kianzarrin** [comment]
> @aubergine10 If two icons hovered at same time, please highlight all affected lanes not just lanes for one icon.

this is more like a bug (you view it as a trick). In any case I increased the distance of the sprites so they do not overlap any more.

**Comment by aubergine10** [comment]
I'd suggest `Reset` instead of `Clear` ("Clear" is a bit ambiguous; might give impression the speed limit is being removed rather than reset to whatever its default is).

**Comment by kvakvs** [comment]
Do not bake keys into translation strings on Crowdin, just do `+ "[Del]"` for now and i can deal with it later.

**Comment by kianzarrin** [comment]
@aubergine10 !  I renamed clear to to default. The screenshot are a commit too old.
> @kvakvs Do not bake keys into translation 

Of course . if you look at the code you will notice its not part of the translation.

**Comment by aubergine10** [comment]
Is there a way to constrain the highlight to where the lanes are - currently it goes over pavement and sometimes cuts in to other side of road too:

image

image

**Comment by aubergine10** [comment]
Sometimes it looks a bit weird, eg. when apply to whole road and there's a corner in the road:

image

**Comment by aubergine10** [comment]
2-lane 1-way road, only half road highligted (direction mode):

image

**Comment by aubergine10** [comment]
@kianzarrin Did you get chance to update the direction-mode lane highlights? Everything else in this PR is good to go.

**Comment by kianzarrin** [comment]
I restored the old lane highlighting method. 

On bend nodes I am either going to get this (currently this is what we get):
Cities_ldBlcTMrYh

or this:


Unless if I do some super clever stuff which is a can of worms I do not want to open right now.

**Comment by kianzarrin** [comment]
I tested this on LHT with invert flag on and off and both directions of start/end nodes.

**Comment by aubergine10** [comment]
For the corners, I prefer the gap rather than the weird overlap.

**Comment by aubergine10** [comment]
Wrong translation key on the reset button in speed palette?

image

Other than that, looks good.

**Comment by kianzarrin** [comment]
Screenshot (581)
Crowdin translations are in and i changed `Checkbox:Default` to `Button:Default`

**Pull request merged**